### PR TITLE
Add Emscripten (WebAssembly) support

### PR DESCRIPTION
**Pull request by srcejon**
Hi.

This patch enables support for the Emscripten (WebAssembly) target that is supported within libusb.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -207,6 +211,11 @@ else()
         target_sources(usb-1.0 PRIVATE
             "${LIBUSB_ROOT}/os/openbsd_usb.c"
         )
+    elseif(EMSCRIPTEN)
+        target_sources(usb-1.0 PRIVATE
+            "${LIBUSB_ROOT}/os/emscripten_webusb.cpp"
+        )
+        target_compile_options(usb-1.0 PRIVATE -pthread)
```
Comment (Youw) [inline CMakeLists.txt:218]:
```suggestion
        find_package(Threads REQUIRED)
        target_link_libraries(usb-1.0 PRIVATE Threads::Threads)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -17,6 +17,10 @@ project(usb-1.0
     VERSION ${LIBUSB_VERSION}
     LANGUAGES C
 )
+if(EMSCRIPTEN)
+    enable_language(CXX)
+    set(CMAKE_CXX_STANDARD 20)
```
Comment (Youw) [inline CMakeLists.txt:22]: isn't `CMAKE_CXX_STANDARD` should be set before the `enable_language(CXX)` to be respected properly?

**Review by Youw — COMMENTED** [review]
Thanks for the contribution.
Is it possible to have a Github Actions updated to have this tested?

File: CMakeLists.txt (hunk 1 above)
Comment (srcejon) [reply CMakeLists.txt:218]: Just adding thread link library with Emscripten isn't sufficient, the compiler also needs the -pthread option.

Without it, we get an error:

    atomic.h:225:10: error: '__builtin_wasm_memory_atomic_notify' needs target feature atomics
      225 |   return __builtin_wasm_memory_atomic_notify((int*)addr, count);

Comment (Youw) [reply CMakeLists.txt:218]: `Threads::Threads` should handle that
is it not?

File: CMakeLists.txt (hunk 2 above)
Comment (Youw) [reply CMakeLists.txt:22]: `enable_language` may create some (temporary?) targets, to verify that the compiler is capable of current selected CXX_STANDARD

File: .github/workflows/builds.yml [hunk 3]
```diff
@@ -264,6 +264,34 @@ jobs:
...
+      run: emcc -v
+    - name: Install dependencies
+      run: |
+        sudo apt update
+        sudo apt install ninja-build
+    - name: Configure
+      run: |
+        rm -rf build install
+        echo "EMSDK:" $EMSDK
+        cmake -B build/static -S src -GNinja -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=install/static -DLIBUSB_BUILD_EXAMPLES=OFF -DLIBUSB_BUILD_TESTING=OFF -DLIBUSB_BUILD_SHARED_LIBS=OFF -DCMAKE_TOOLCHAIN_FILE=$EMSDK/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake
```
[the remark this answers is not in the document]
Comment (srcejon) [reply .github/workflows/builds.yml:285]: Emscripten doesn't support shared libraries. (If you try to build a shared library, the tools will actually build a static library)

**Review by Youw — APPROVED** [review]

**Comment by Youw** [comment]
Thanks for the contribution.
I've cherry-picked this to 1.0.27 branch as well.

**Pull request merged**